components/script/dom/element.rs

```diff
@@ -534,13 +537,43 @@ impl Element {
             }
         }
 
-        // TODO: Update the following steps to align with the newer spec.
         // Step 3.
+        // If element’s local name is a valid custom element name,
+        // or element’s is value is non-null
+        if is_valid_custom_element_name(self.local_name()) || self.get_is().is_some() {
+            let definition = self.get_custom_element_definition();
+            if definition.is_some() && definition.unwrap().disable_shadow {
+                return Err(Error::NotSupported);
+            }
+        }
+
+        // Step 4. If element is a shadow host:
         if self.is_shadow_host() {
-            return Err(Error::InvalidState);
+            let current_shadow_host = self.shadow_root().unwrap();
```

Comment:
You can avoid calling unwrap here by doing something like this:
```rust
// Step 4. If element is a shadow host:
if let Some(shadow_root) = self.shadow_root() {
```

And please don't refer to the shadow root as a shadow host, these are seperate concepts. The element that a shadow root is attached to is a shadow host.